lib/Runtime/ByteCode/ScopeInfo.cpp

```diff
@@ -162,23 +162,20 @@ namespace Js
         Scope* currentScope = byteCodeGenerator->GetCurrentScope();
         Assert(currentScope->GetFunc() == funcInfo);
 
-// Disable the below temporarily because of bad interaction with named function expressions
-#if 0
         if (funcInfo->root->IsDeclaredInParamScope()) {
             Assert(currentScope->GetScopeType() == ScopeType_FunctionBody);
             Assert(currentScope->GetEnclosingScope());
 
             FuncInfo* func = currentScope->GetEnclosingScope()->GetFunc();
             Assert(func);
 
```

Comment:
Paging this back in... So "currentScope" in this case is the body scope of the function associated with "funcInfo". Is that right? And what we're trying to do by accessing currentScope->GetEnclosingScope()->GetFunc() is get the enclosing function?

Reply:
`currentScope` is the body scope for the function which is declared in a parameter scope. `currentScope->GetEnclosingScope()->GetFunc()` is supposed to get us the function enclosing it - the function owning the parameter scope in which the nested function is located. Does that sound right? I hope I put the words together correctly here because it's kind of subtle.

Reply:
Okay. That's what I thought. Are we guaranteed to get the enclosing function that way? If the "currentScope" function has a param scope or func expr scope enclosing the body scope, then that's what GetEnclosingScope() will give us.

Reply:
We can keep looking at enclosing scopes while the function is the same as `currentScope->GetEnclosingScope()->GetFunc()` to skip any other scopes enclosing the body scope, right?

I tried to find a case which hits that but didn't see one. Here's an example which doesn't need the above mitigation - `currentScope->GetEnclosingScope()->GetFunc()` returns the enclosing function for the function defined in the param scope.

```javascript
var foo5 = function foo5a(a = (function(){(function(b = 123) { +x; })()})()) {
    function bar() { eval(''); }
    var x;
};

foo5();
```

Could you help me build a case to test this out?

Reply:
I don't have a test case off the top of my head, but I wonder if we could use the FuncInfoStack instead of the scope stack to find the enclosing func.

Reply:
I don't know of a method which uses the funcInfoStack to get anything other than the Top item. Can add a helper which does something like this:
```c++
FuncInfo* ByteCodeGenerator::GetEnclosingFuncInfo()
{
    FuncInfo* top = this->funcInfoStack->Pop();

    Assert(!this->funcInfoStack->Empty());

    FuncInfo* second = this->funcInfoStack->Top();

    this->funcInfoStack->Push(top);

    return second;
}
```
We already have a helper which works instead of manually walking the enclosing scope chain. It's not exposed outside ByteCodeGenerator.cpp but could be:
```c++
FuncInfo* ByteCodeGenerator::GetParentFuncInfo(FuncInfo* child)
{
    for (Scope* scope = child->GetBodyScope(); scope; scope = scope->GetEnclosingScope())
    {
        if (scope->GetFunc() != child)
        {
            return scope->GetFunc();
        }
    }
    Assert(0);
    return nullptr;
}
```
Interesting. I used the first method above to get the enclosing function via the funcInfoStack but hit some unit test failures. The stack looks to be wrong in some cases when the enclosing function is split-scope. Here's an example:
```javascript
function foo(a, b = function() {a;})
{
...
}
```
In that case, the funcInfoStack does not insert foo between the global and anonymous function in the param scope. Using the first helper above, we end up returning the global function. Using the helper which walks up the enclosing scopes works correctly, though.

Reply:
Urgh. Can you figure out where the FuncInfo stack got screwed up? I'd hate to leave broken logic in there once we've discovered it.

Reply:
Yeah, I'm looking into that right now. Finding that the stack was just wrong was very surprising.

Reply:
Looks like when we start emitting the function with a split scope, we'll emit the param scope before pushing the FuncInfo for the enclosing function. That means functions defined in the param scope will see the wrong FuncInfoStack.

I pushed an update to this PR which corrects the stack in that case. Seems to be more correct and the `GetEnclosingFuncInfo` function above works now.

Reply:
Awesome. Thanks. I’ll look tomorrow.